### ISSUE TITLE
Feature request: warn on set creation with unhashable types

### ISSUE THREAD

**Issue by gtkacz** [astral-sh/ruff#21349]
### Summary

If an user is trying to create a `set` or similar on unhashable types, it will error with something like `TypeError: unhashable type: 'list'`. This can easily be warned by `ruff` if it knows the type of the variable by checking for the existence of the `__hash__` method. 

Which would warn the user in something like:

```py
some_set = set([{'foo': 'bar'}])
```

Worth pointing out that iterables with root-level hashable types can be passed, e.g.:

```py
set({'foo': 'bar'})
>>> {'foo'}

set([0, 1, 2])
>>> {0, 1, 2}
```

Example checker implementation in Python:

```py
def is_unhashable(var: Any) -> bool:
	if not var.__class__.__hash__ and (var.__class__.__iter__ and not var[0].__class__.__hash__):
		return True

	return False

is_unhashable([0, 1, 2])
>>> False

is_unhashable([{'foo': 'bar'}])
>>> True

is_unhashable('foo')
>>> True
```

**Comment by ntBre** [comment]
My first thought was that this might be a better fit for a type checker, but it doesn't look like [ty](https://play.ty.dev/1f0224b0-f593-40d3-a1d4-a59cf0b6eada), [pyright](https://pyright-play.net/?pythonVersion=3.8&strict=true&code=M4ewtgpg%2BsEC4AIC8DZwBQG0DeByAZiCLgFwK4BGAhgE64C%2BAugJRA), or [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=8ce24737929eb07d09bd82545ada1a42) flag this example either.

**Comment by MichaReiser** [comment]
Yeah, that was my thought too but `set` takes an `Iterable[T]` without any constraints on `T`. There's an `Hashable` Protocol but it isn't widely used. @AlexWaygood probably knows more. 

We could hard code common cases but a proper implementation requires multifile analysis, which Ruff doesn't support yet.

**Comment by dylwil3** [comment]
A relevant `typeshed` issue: https://github.com/python/typeshed/issues/9571

**Comment by AlexWaygood** [comment]
Type checkers are mostly useless for accurately inferring hashability, unfortunately. There is a `Hashable` protocol, but a major problem is that `object` is inferred as a subtype of `Hashable` (because `hash(object())` works at runtime). You can logically infer from this that all Python objects are hashable, since all Python types are subtypes of `object`... but wait, that's obviously not true.

Hashability in Python just doesn't follow the normal rules that Python type checkers are meant to assume when they type-check Python code; it's a fundamental violation of the Liskov Substitution Principle baked deep into the language. I actually think Ruff may be able to do better than ty here just by hardcoding a few common cases of things that are known not to be hashable; it's very hard for type checkers to get this right.

**Comment by AlexWaygood** [comment]
For related discussion, see:
- https://github.com/astral-sh/ty/issues/1132
- https://github.com/astral-sh/ruff/pull/20284
- https://github.com/PyCQA/flake8-pyi/issues/283#issuecomment-1248095891
- https://github.com/astral-sh/ruff/blob/1d188476b608c94cea5cb2a97ae248d70a0e2154/crates/ty_vendored/vendor/typeshed/stdlib/typing.pyi#L867-L873

**Comment by gtkacz** [comment]
@AlexWaygood should I close this then? Looks like it's a better fit for `ty`
